libraries/ts-command-line/src/CommandLineParameterProvider.ts

```diff
@@ -82,97 +144,102 @@ abstract class CommandLineParameterProvider {
   }
 
   /**
-   * Defines a command-line parameter whose value must be a string from a fixed set of
-   * allowable choice (similar to an enum).
-   *
-   * @remarks
-   * Example:  example-tool --log-level warn
+   * Generates the command-line help text.
    */
-  protected defineOptionParameter(definition: ICommandLineOptionDefinition): CommandLineOptionParameter {
-    if (!definition.options) {
-      throw new Error(`When defining an option parameter, the options array must be defined.`);
-    }
-    if (definition.defaultValue && definition.options.indexOf(definition.defaultValue) === -1) {
-      throw new Error(`Could not find default value "${definition.defaultValue}" `
-        + `in the array of available options: ${definition.options.toString()}`);
-    }
-    return this._createParameter(definition, {
-      choices: definition.options,
-      defaultValue: definition.defaultValue
-    }) as CommandLineOptionParameter;
+  public renderHelpText(): string {
+    return this._getArgumentParser().formatHelp();
   }
 
+  /**
+   * The child class should implement this hook to define its command-line parameters,
+   * e.g. by calling defineFlagParameter().
+   */
+  protected abstract onDefineParameters(): void;
+
+  /**
+   * Retrieves the argparse object.
+   * @internal
+   */
+  protected abstract _getArgumentParser(): argparse.ArgumentParser;
+
   /** @internal */
   protected _processParsedData(data: ICommandLineParserData): void {
     // Fill in the values for the parameters
     for (const parameter of this._parameters) {
-      parameter._setValue(data);
+      const value: any = data[parameter._parserKey];
+      parameter._setValue(value);
     }
   }
 
-  private _getKey(
-    parameterLongName: string,
-    key: string = 'key_' + (CommandLineParameterProvider._keyCounter++).toString()): string {
+  private _generateKey(): string {
+    return 'key_' + (CommandLineParameterProvider._keyCounter++).toString();
+  }
 
-    const existingKey: string | undefined = this._keys.get(key);
-    if (existingKey) {
-      throw colors.red(`The parameter "${parameterLongName}" tried to define a key which was already ` +
-        `defined by the "${existingKey}" parameter. Ensure that the keys values are unique.`);
-    }
+  private _getFlagParameter<T extends CommandLineParameter<any>>(parameterLongName: string,
+    expectedKind: CommandLineParameterKind): T {
 
-    this._keys.set(key, parameterLongName);
-    return key;
+    const parameter: CommandLineParameter<any> | undefined = this._parametersByLongName.get(parameterLongName);
+    if (!parameter) {
+      throw new Error(`The parameter "${parameterLongName}" is not defined`);
+    }
+    if (parameter.kind !== expectedKind) {
+      throw new Error(`The parameter "${parameterLongName}" is of type "${CommandLineParameterKind[parameter.kind]}"`
+        + ` whereas the caller was expecting "${CommandLineParameterKind[expectedKind]}".`);
+    }
+    return parameter as T;
   }
```

Comment:
> } [](start = 1, length = 2)

You should read all the parameter values upfront. Otherwise if you rename an expected value you may not catch the mistake because it is a runtime error off in some conditional block somewhere.

Reply:
We'll put this on the web site

---
In reply to: [182196299](https://github.com/Microsoft/web-build-tools/pull/611#discussion_r182196299) [](ancestors = 182196299)